Source/Orts.Simulation/Simulation/Physics/Train.cs

```diff
@@ -14097,13 +14099,315 @@ public String[] AddRestartTime(String[] stateString)
+                                PlayerTrainDivergingSwitches[dir, 1].Add(thisItem);
+                            }
+                        }
+                    }
+                    // search for mileposts
+                    if (thisSection.CircuitItems.TrackCircuitMileposts != null)
+                    {
+                        foreach (TrackCircuitMilepost thisMilepostItem in thisSection.CircuitItems.TrackCircuitMileposts)
+                        {
+                            Milepost thisMilepost = thisMilepostItem.MilepostRef;
+                            var distanceToTrainM = thisMilepostItem.MilepostLocation[sectionDirection == 1 ? 0 : 1] + sectionDistanceToTrainM;
+                            if (distanceToTrainM < maxDistanceM)
+                            {
+                                if (!(distanceToTrainM - prevMilepostDistance < 50 && thisMilepost.MilepostValue == prevMilepostValue) && distanceToTrainM > 0)
+                                {
+                                    thisItem = new TrainObjectItem(thisMilepost.MilepostValue.ToString(), distanceToTrainM);
+                                    prevMilepostDistance = distanceToTrainM;
+                                    prevMilepostValue = thisMilepost.MilepostValue;
+                                    PlayerTrainMileposts[dir].Add(thisItem);
+                                }
+                            }
+                            else break;
+                        }
+                    }
+                    // search for tunnels
+                    if (thisSection.TunnelInfo != null)
+                    {
+                        foreach (var thisTunnel in thisSection.TunnelInfo)
+                        {
+                            var tunnelStartOffset = thisTunnel[sectionDirection].TunnelStart;
+                            var distanceToTrainM = tunnelStartOffset + sectionDistanceToTrainM;
+                            if (distanceToTrainM < maxDistanceM)
+                            {
+                                if (tunnelStartOffset > lengthOffset)
+                                {
+                                    thisItem = new TrainObjectItem(tunnelStartOffset + sectionDistanceToTrainM, (int)thisTunnel[sectionDirection].TotalLength, TrainObjectItem.TRAINOBJECTTYPE.TUNNEL);
+                                    PlayerTrainTunnels[dir].Add(thisItem);
+                                }
+                                else if (PlayerTrainTunnels[dir].Count == 0 && (thisTunnel[sectionDirection].TunnelEnd < 0 || thisTunnel[sectionDirection].TunnelEnd > lengthOffset))
+                                {
+                                    // Train is in tunnel, compute remaining length
+                                    var remainingLength = thisTunnel[sectionDirection].TotalLength - lengthOffset - (tunnelStartOffset < 0 ? thisTunnel[sectionDirection].TCSStartOffset : tunnelStartOffset);
```

Comment:
I made a mistake in the calculation of the remaining length of the tunnel. It should be:
``` csharp
var remainingLength = thisTunnel[sectionDirection].TotalLength - lengthOffset + (tunnelStartOffset < 0 ? -thisTunnel[sectionDirection].TCSStartOffset : tunnelStartOffset);
```
(There was a missing minus sign)

Could you please insert the modification?

Reply:
OK, corrected.